documentation/source/protocol_extension.rst

```diff
@@ -0,0 +1,90 @@
+How to write a protocol extension
+=================================
+
+Cocotb take care of the core communication between the simulator and testbench.
+To use some known busses like wishbone, avalon, spi, i2c,... An extension is required.
+
+This is a guideline that will explain how to write a python module protocol
+extension for cocotb.
+
+Naming convention
+-----------------
+
+A cocotb extension module must begin by `cocotb_` and terminate with protocol
+name.
+
+For exemple, with spi the module extension is `cocotb_spi`.
+
+Bus driver
+----------
+
+A Bus driver is a python class that factorize common behavior and check for
+presence of some signals.
+
+A Bus driver is an active class that be driven in the main testbench. Example is
+given with avalon bus in `cocotb/drivers/avalon.py`.
+
+With AvalonMaster class we can generate stimulis for reading and writing on bus.
+
+The main class is named `BusDriver` and is defined in `cocotb/drivers/__init__.py`
+
+TODO: give an example.
+
+Bus Monitor
+-----------
+
+A Bus monitor is mainly a passive class that will monitor the bus.
+
+TODO: explain monitor
+
+Python module
+-------------
+
+The module file hierarchy should be as following ::
+
+  ├── cocotb_spi/
+  │   └── __init__.py
+  ├── README.md
+  └── setup.py
+
+With python module source code in directory cocotb_spi and module configuration
+in setup.py
+
+setup.py is a python configuration file that be used for packaging. An example
+is given above ::
+
+  try:
+    from setuptools import setup
+  except ImportError:
+    from distutils.core import setup
```

Comment:
Does cocotb do this? If not, we may as well assume setuptools

Reply:
As I understand setuptools, the package `distutils.core` is for old python version (python2) and new python version use from `setuptools`.

Reply:
Make sure the ``setup.py`` contains lines like these two: https://github.com/wallento/cocomod-uart/blob/64f7af8922ea67efda4a1a3c185ce2ecaa45e413/setup.py#L18 and https://github.com/wallento/cocomod-uart/blob/64f7af8922ea67efda4a1a3c185ce2ecaa45e413/setup.py#L26.

Reply:
I discover the module cocomod-uart. Should not it be referenced in github [cocotb organisation ](https://github.com/cocotb)?

Reply:
> I discover the module cocomod-uart. Should not it be referenced in github [cocotb organisation ](https://github.com/cocotb)?

It's already kinda tied in by this new "Used by:" button that GitHub shows in the top row that leads to https://github.com/cocotb/cocotb/network/dependents